sim/core/rppm.go

```diff
@@ -0,0 +1,321 @@
+		return 1 / unit.CastSpeed
+	case HighestHaste:
+		// as of 5.2 this should no longer include non 'True haste mods' so only i.E. Lust
+		// With that this all should be equal
+		return unit.PseudoStats.AttackSpeedMultiplier * (1 + unit.GetStat(stats.HasteRating)/(HasteRatingPerHastePercent*100))
+	case LowestHaste:
+		return math.Min(unit.CastSpeed, math.Min(unit.SwingSpeed(), unit.RangedSwingSpeed()))
+	default:
+		return 1.0
+	}
+}
+
+type rppmSpecMod struct {
+	spec        proto.Spec
+	coefficient float64
+}
+
+func (r rppmSpecMod) GetCoefficient(charater *Character) float64 {
+	if charater.Spec == r.spec {
+		return 1 + r.coefficient
+	}
+
+	return 1.0
+}
+
+type rppmClassMod struct {
+	classMask   int
+	coefficient float64
+}
+
+func toDBCClass(class proto.Class) int {
+	switch class {
+	case proto.Class_ClassWarrior:
+		return 1
+	case proto.Class_ClassPaladin:
+		return 2
+	case proto.Class_ClassHunter:
+		return 3
+	case proto.Class_ClassRogue:
+		return 4
+	case proto.Class_ClassPriest:
+		return 5
+	case proto.Class_ClassDeathKnight:
+		return 6
+	case proto.Class_ClassShaman:
+		return 7
+	case proto.Class_ClassMage:
+		return 8
+	case proto.Class_ClassWarlock:
+		return 9
+	case proto.Class_ClassMonk:
+		return 10
+	case proto.Class_ClassDruid:
+		return 11
+	default:
+		return 1
+	}
+}
+
+func (r rppmClassMod) GetCoefficient(character *Character) float64 {
+	mask := 1 << (toDBCClass(character.Class) - 1)
+	if r.classMask&mask > 0 {
+		return 1 + r.coefficient
+	}
+
+	return 1
+}
+
+type rppmApproxIlvlMod struct {
+	baseIlvl    int32
+	realIlvl    int32
+	coefficient float64
+}
+
+func (r rppmApproxIlvlMod) GetCoefficient(unit *Unit) float64 {
+
+	// We use an approximation here, or we'd need to load the complete random properties table into the sim
+	// Just to calculate the difference in random prop points as not all points are available on the item that scales
+	// The maximal relativ error I observed when comparing real values and approximation was 0.07%
+	// which was an increase of 0,0008% in proc chance. So for now I think we can neglect that
+
+	if r.baseIlvl == r.realIlvl {
+		return 1
+	}
+
+	// Each ilvl step is ~0.936% larger than the previous. They're rounded in the table
+	// But over a range of >3 ilvl this approximation becomes very accurate
+	return 1 + (math.Pow(1.00936, float64(r.realIlvl-r.baseIlvl))-1)*r.coefficient
+}
+
+type HasteRPPMModKind uint8
+
+const (
+	MeleeHaste HasteRPPMModKind = iota
+	RangedHaste
+	SpellHaste
+	HighestHaste
+	LowestHaste
+)
+
+type CritRPPMModKind uint8
+
+const (
+	MeleeCrit CritRPPMModKind = iota
+	RangedCrit
+	SpellCrit
+	LowestCrit
+)
+
+type RPPMProc struct {
+	ppm             float64
+	coefficient     float64
+	charCoefficient float64
+	char            *Character
+	lastProc        time.Duration
+	lastCheck       time.Duration
+	mod             []rppmMod
+	charMods        []rppmCharMod
+}
+
+// Attach a crit mot to the RPPM Proc
+// The most common value used is LowestCrit
```

Comment:
Are you sure that the most common is `LowestCrit` rather than `HighestCrit`? Seems odd that it would scale up a caster's RPPM proc chance using their physical Crit chance.